### DISCUSSION TITLE
Name lookup service in Home Assistant - switch to systemd-resolved

### DISCUSSION THREAD

**Discussion by NODeeJay** [home-assistant/architecture#768] [General]
Hi!

coming over from that conversation https://community.home-assistant.io/t/local-dns/178108/145 with Mike I have a RFC (Request for conversation):

### Summary:
The name resolution service in Home Assistant in a supervised installation was part of many discussions in the forum and base for some PRs. Additionally it caused [CVE-2020-36517](https://nvd.nist.gov/vuln/detail/CVE-2020-36517). Currently there are 2 name resolution serives in Home Assistant supervised: _systemd-resolved_ and _coreDNS_, both need maintenance. This situation is not optimal, I suggest conversation about a possible one-stop solution and a configuration that respects IETF/RFC/IANA considerations and provides the best user experience in terms of flexibility vs. usability vs. complexity vs. security.
My understanding of HA architecture is limited and it is even worse with Python, I may not see a point that is super clear for you, I can see it from my perspective as a user and seasoned system administrator working in industrial IoT.

### in detail - requirements from my POV
- MUST resolve names for the internal docker network
- MUST resolve mDNS
- MUST resole LLMNR
- MUST resolve DNS with ANY variant: DoT/DNS - is DoH needed when we have DoT?
- MUST respect DHCP assigned connection specific search domains and statically configured search domains (static IP)
- OPTIONAL resolve SSDP (haven't seen anywhere in HA, is it needed?)

### in detail - possible solution: switch entirely to systemd-resolved
- it supports all existing configurations from coreDNS including a fallback option
- only one name resolution service is easier to maintain
- no separate container for coreDNS needed (saves memory, disk, CPU, code)

### in detail - with this config (and why)
/etc/systemd/resolved.conf
```
[Resolve]
#DNS=
#FallbackDNS=1.1.1.1 8.8.8.8 1.0.0.1 8.8.4.4 2606:4700:4700::1111 2001:4860:4860::8888 2606:4700:4700::1001 2001:4860:4860::8844
#Domains=
DNSSEC=allow-downgrade #use DNSSEC if possible but allow downgrade
DNSOverTLS=opportunistic #use DoT if possible but allow downgrade
MulticastDNS=resolve/true #to handle mDNS
LLMNR=resolve/true #to handle LLMNR
#Cache=yes
DNSStubListener=no # this is maybe the solution to offer DNS service for the internal docker network?
#ReadEtcHosts=yes
ResolveUnicastSingleLabel=no #disable to prevent forward A or AAAA singel-label domains to DNS servers 
```

In a setup with static IP the nameserver(s) is/are populated into /etc/resolve.conf, additionally the search domains need to be populated - actually that happens already when the interface is set to DHCP and a connection specific suffix is part of the DHCP offer, but not when a static IP is set - so here it's just a cleanup of an inconsistency, see https://github.com/home-assistant/operating-system/issues/1916#issue-1239296283:
```
# Generated by NetworkManager
nameserver 192.168.100.32
nameserver 192.168.100.33
search mydomain.bla
search myotherdomain.bla
```

Fallback is not needed anymore, it is included already, from(https://wiki.archlinux.org/title/Systemd-resolved#Fallback):
If systemd-resolved does not receive DNS server addresses from the [network manager](https://wiki.archlinux.org/title/Network_manager) and no DNS servers are configured [manually](https://wiki.archlinux.org/title/Systemd-resolved#Manually) then systemd-resolved falls back to the fallback DNS addresses to ensure that DNS resolution always works.
Note: The fallback DNS are in this order: [Cloudflare](https://wiki.archlinux.org/title/Alternative_DNS_services#Cloudflare), [Quad9](https://wiki.archlinux.org/title/Alternative_DNS_services#Quad9) (without filtering and without DNSSEC) and [Google](https://wiki.archlinux.org/title/Alternative_DNS_services#Google); see the [systemd PKGBUILD](https://github.com/archlinux/svntogit-packages/blob/packages/systemd/trunk/PKGBUILD#L100-L111) where the servers are defined.

As far as I was able to verify currently SSDP is not used/supported in HA?

This change is from my POV a major change in handling name resolution, it needs extensive testing.
I can contribute tests in a heterogenous routed mydomain.local network with Win/Apple/Android where DNS, mDNS and SSDP is used (LLMNR is disabled). 

### Test cases:
- all lookups to be performed in the core container and one add on container (e.g. studio code server)
- lookups against FQDN, LLMNR or mDNS :
  - WAN address, e.g. www.google.com (FQDN)
  - local internal lookup within the docker network
  - local lookups against
    - IF DHCP: the connection specific search domain (FQDN) e.g. mypc.mydomain.local or mypc.mydomain.home
    - IF static: the user's provided search domains in his configured order (FQDN) e.g. mypc.mydomain.local or mypc.mydomain.home
    - mDNS e.g. mymac.local
    - LLMNR - and I have no clue how to test that...

### The possible mess
From a support POV: incorrect DNS or search domain settings will result in serious migrane attacks! resolved does not allow scripting like coreDNS does, the existing fallback feature cannot be ported. Multiple individual DNS server configurations on different ports or interfaces do not exist in resolved, DNSStubListener and DNSStubListenerExtra allow providing the same options on another interface, or just a forwarder, but no individual configuration, however, they may be a solution for the internal docker network.
From a users perspective there are beginner and advanced settings. DHCP is beginner - zeroconf. While someone who wants to assign a static IP, a search domain and thus creating a FQDN also for HA is surely not beginner and will hopefully know what he/she does. Why not hiding static IP, netmask, gateway, DNS servers and search domains behind the advanced user settings?

And what I don't know how many add-ons and functions will break, but I am available for thouroug testing.

What do you guys think about it?

Cheers,
Alex

**Comment by shallax** [comment]
It’s maddening that Home Assistant doesn’t allow me to specify a .local domain in DNS. 

The most recent upgrade broke all of my installation - couldn’t connect to MySQL, couldn’t find a bunch of my devices, lost connection to my Zwave controller - all because they’re {hostname}.local on my network. 

Why can every other OS out there support mdns and still allow you to assign .local domain names to devices but Home Assistant can’t? 

This really needs to be fixed.

**Reply by NODeeJay** [reply]
Hi Shallax,

I so understand and share your frustration, see: https://community.home-assistant.io/t/local-dns/178108/142

You write "they're {hostname}.local". Do they form a proper FQND with 3 labels or just 2 labels. mypc.local vs. mypc.mydomain.local
The latter will work because one service can respond to .local (mDNS) properly and the other service responds to mydomain.local (DNS).
But when you just have mypc.local it won't work in the way DNS and mDNS is built.
Also the coexistence of DNS with a mydomain.local and mDNS will not resolve one possible hostname from mDNS through mDNS but as the nameless A record for said domain: mydomain.local

There are a bunch of stones in our way, starting with the 2 competing DNS services. There are options to resolve. In first instance I started digging into coreDNS and add the local domain there as a forward, which would make coreDNS answer for that local domain first, see here: https://community.home-assistant.io/t/local-dns/178108/144

But this is most probably not the solution and will make things worse. Why. Because there is systemd-resolved which is also used. So the 2 DNS would be even more competing.

Mike mentioned that there seem to be a certain interest to switch to resolved as a one-stop solution but it seems that it did not went hassle free in their tests.

If you want your issue to be fixed now, you have some options, also please see (https://github.com/home-assistant/operating-system/issues/1916):
1. configure your search domain either by 
  2. pushing it with your DHCP to HA
  3. setting it with resolvectl INTERFACE domain MYDOMAIN.LOCAL
  4. Check the thread above and what I wrote in the forum with Mike it will quickly give you clue how to workaround it.

**Reply by shallax** [reply]
Yeah, unfortunately I just had stuff as "{hostname}.local" with just 2 labels. Everything worked fine up until now.
